sticker-utils/src/bin/sticker-train.rs

```diff
@@ -118,113 +176,59 @@ fn train_model(
     Ok(())
 }
 
-fn run_epoch(
+fn run_epoch<E>(
+    config: &Config,
+    encoder: &mut CategoricalEncoder<E, E::Encoding>,
+    vectorizer: &SentVectorizer,
     trainer: &TaggerTrainer,
-    tensors: &CollectedTensors,
+    file: &mut File,
     is_training: bool,
     lr: f32,
-) -> (f32, f32) {
+) -> (f32, f32)
+where
+    E: SentenceEncoder,
+    E::Encoding: Clone + Eq + Hash,
+{
     let epoch_type = if is_training { "train" } else { "validation" };
 
+    let read_progress = ReadProgress::new(file).or_exit("Cannot create progress bar", 1);
+    let progress_bar = read_progress.progress_bar().clone();
+    progress_bar.set_style(ProgressStyle::default_bar().template(&format!(
+        "[Time: {{elapsed_precise}}, ETA: {{eta_precise}}] {{bar}} {{percent}}% {} {{msg}}",
+        epoch_type
+    )));
+
+    let mut dataset = ConllxDataSet::new(read_progress);
+
     let mut instances = 0;
     let mut acc = 0f32;
     let mut loss = 0f32;
 
-    let progress = ProgressBar::new(tensors.labels.len() as u64);
-    progress.set_style(ProgressStyle::default_bar().template(&format!(
-        "{{bar}} {} batch {{pos}}/{{len}}, {{msg}}",
-        epoch_type
-    )));
-
-    for i in 0..tensors.labels.len() {
-        let seq_lens = &tensors.sequence_lens[i];
-        let tokens = &tensors.inputs[i];
-        let labels = &tensors.labels[i];
+    for batch in dataset
+        .batches(encoder, vectorizer, config.model.batch_size)
+        .or_exit("Cannot read batches", 1)
+    {
+        let (inputs, seq_lens, labels) = batch.or_exit("Cannot read batch", 1).into_parts();
 
         let batch_perf = if is_training {
-            trainer.train(seq_lens, tokens, labels, lr)
+            trainer.train(&seq_lens, &inputs, &labels.0, lr)
```

Comment:
Maybe using a named field rather than `.0` could make this clearer?
Or define a function on `LabelTensor` like `tensor()` returning a ref to the inner tensor as `labels.tensor()`?

Reply:
I'd be in favor of a named field. But I guess that this should be a separate change. Maybe we don't need the `LabelTensor` wrapper at all, I guess we could just directly implement the `Labels` trait for `NdTensor<i32, Ix2>`. I'll do a separate PR for this.